Source/WebCore/css/ImmutableStyleProperties.cpp

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1.  Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ * 2.  Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ImmutableStyleProperties.h"
+
+#include "CSSCustomPropertyValue.h"
+#include "StylePropertiesInlines.h"
+
+namespace WebCore {
+
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(ImmutableStyleProperties);
+
+ImmutableStyleProperties::ImmutableStyleProperties(const CSSProperty* properties, unsigned length, CSSParserMode mode)
+    : StyleProperties(mode, length)
+{
+    auto* metadataArray = const_cast<StylePropertyMetadata*>(this->metadataArray());
+    auto* valueArray = bitwise_cast<PackedPtr<CSSValue>*>(this->valueArray());
+    for (unsigned i = 0; i < length; ++i) {
```

Comment:
Man, a ZipperedIterator would be nice here:

```c++
     auto* metadataArray = const_cast<StylePropertyMetadata*>(this->metadataArray());
     auto* valueArray = bitwise_cast<PackedPtr<CSSValue>*>(this->valueArray());
    for (auto& [metadata, value, property] : WTF::zippered(metadataArray, valueArray, properties)) {
         metadata = property.metadata();
         value = property.value();
         value->ref();
    }
```

Maybe it's time to add one. Seems like a fun weekend project for me - https://bugs.webkit.org/show_bug.cgi?id=251319.

Reply:
Yes, this would work if we switch from pointers to spans. Might have to rejigger the `const` handling too.